File: topics/stacks/stack.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

#define MAX 10

bool isFull(int* top) {
    return *top == MAX - 1;
}

bool isEmpty(int* top) {
    return *top == -1;
}

void push(int *top, int stacks[], int ele) {
    if (isFull(top)) {
        printf("Stack is Full! Cannot Push.\n");
        return;
    }
    stacks[++(*top)] = ele;
}

int pop(int *top, int stack[]) {
    if (isEmpty(top)) {
        printf("Stack is empty! Cannot Pop\n");
        return -1;
    }
    return stack[(*top)--];
}

int peek(int* top, int stack[]) {
    if (isEmpty(top)) {
        printf("Stack is Empty! Cannot Peek\n");
        return -1;
    }
    return stack[*top];
}
void display(int *top, int stack[]) {
    if (isEmpty(top)) {
        printf("Stack is empty!\n");
        return;
    }
    printf("Stack contents: ");
    for (int i = 0; i <= *top; i++) {
        printf("%d ", stack[i]);
    }
    printf("\n");
}


int main() {
    int stacks[MAX];
    int top = -1;
    int ch = 0;
    int ele = 0;

    printf("Stacks\n");
    printf("1. Push \n2. Pop \n3. Peek \n4. isEmpty \n5. isFull \n6. Display \n7. Exit \n");

    while (1) {
        printf("\nEnter your choice: ");
        scanf("%d", &ch);

        switch (ch) {
            case 1:
                printf("Enter Element: ");
                scanf("%d", &ele);
                push(&top, stacks, ele);
                break;
            case 2:
                printf("Popped: %d\n", pop(&top, stacks));
                break;
            case 3:
                printf("Top Element: %d\n", peek(&top, stacks));
                break;
            case 4:
                printf(isEmpty(&top) ? "Stack is Empty\n" : "Stack is Not Empty\n");
                break;
            case 5:
                printf(isFull(&top) ? "Stack is Full\n" : "Stack is Not Full\n");
                break;
            case 6:
                display(&top,stacks);
                break;
            case 7:
                printf("Exiting\n");
                return 0;
            default:
                printf("Invalid Input\n");
        }
    }
}

File: topics/stacks/stack.py
class Stacks:
    def __init__(self, max=5):
        self.stack = []          
        self.max = max           

    def isFull(self):
        return len(self.stack) == self.max

    def isEmpty(self):
        return len(self.stack) == 0

    def push(self, ele):
        if self.isFull():
            print("Stack is Full. Cannot Push.")
        else:
            self.stack.append(ele)
            print(f"Pushed: {ele}")

    def pop(self):
        if self.isEmpty():
            print("Stack is Empty. Cannot Pop.")
            return None
        return self.stack.pop()

    def peek(self):
        if self.isEmpty():
            print("Stack is Empty. Cannot Peek.")
        else:
            print(f"Top Element: {self.stack[-1]}")

    def display(self):
        if self.isEmpty():
            print("Stack is Empty. Nothing to display.")
        else:
            print("Stack Contents:", *self.stack)


if __name__ == "__main__":  
    my_stack = Stacks(5)    

    while True:
        print("\n1. Push\n2. Pop\n3. Peek\n4. isEmpty\n5. isFull\n6. Display\n7. Exit")
        choice = input("Enter your choice: ")

        if choice == "1":
            value = input("Enter element to push: ")
            my_stack.push(value)

        elif choice == "2":
            popped_value = my_stack.pop()
            if popped_value is not None:
                print(f"Popped: {popped_value}")

        elif choice == "3":
            my_stack.peek()

        elif choice == "4":
            print("Stack is empty!" if my_stack.isEmpty() else "Stack is NOT empty.")

        elif choice == "5":
            print("Stack is full!" if my_stack.isFull() else "Stack is NOT full.")

        elif choice == "6":
            my_stack.display()

        elif choice == "7":
            print("Exiting...")
            break

        else:
            print("Invalid choice. Try again!")
